### PR TITLE
[1075] Introduce Further Education bodies

### PR DESCRIPTION
**Pull request by asmega**
### Context

- https://trello.com/c/R2235yoT/1075-spike-adding-fe-institutions-to-the-service
- Handle Further Education bodies

### Out of scope

- FE schools will only have one delivery address like any other school we currently have in the service

### Changes proposed in this pull request

- Introduce Single Table Inheritance to `School`
- `School` will be the parent class and abstract class
- All current schools will become `CompulsorySchool`
- Further Education schools will be added as `FurtherEducationSchool`
- Add `FurtherEducationCollection` as a child of `ResponsibleBody`

All Further Education schools will be imported to have a responsible body despite the fact that technically they do not have a responsible body. This is because the there is a lot of code that assumes the fact that a school simply has an RB

- Add `UKPRN` to `schools` table
- NOTE: an FE body must have a UKPRN and may also have a URN too

- As we have use `URN` to identify schools in URLs the look up will now search for for schools via `UKPRN` then URN as schools can now potentially have 2 identifiers

#### CC integration

- As `FurtherEducationSchool` will have an RB (FurtherEducationCollege) the generation of RB and Schools in `/computacenter` remain similar to how current RBs and Schools worl
- `UserChanges` behave the same as any current `School` only not us the RB "urn" and School "urn" identifiers will be the same 

#### Next steps

- Either add copy/content directly to schools controllers namespace OR add another controller namespace called `/further_education_schools` to give custom journey for these users
- Once list of FE bodies are finalised, these can be imported into the service

### Questions

- Where do we attach users to? As we have a placeholder RB and the school they could be on either. My thoughts are they should be added to the RB. My understanding is that RB users according CC have more permissions eg request BIOS resets?

### Guidance to review

n/a

### CODE REVIEW

File: app/form_objects/school_search_form.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def select_order_state_options
 
   def csv_filename
     tokens = %w[allocations]
-    tokens << "#{array_of_urns.count}-URNs" if urns.present?
+    tokens << array_of_identifiers.count.to_s if identifiers.present?
```
Comment (jamesrose) [inline app/form_objects/school_search_form.rb:47]: Are there any issues changing this token format?

Comment (jamesrose) [inline app/form_objects/school_search_form.rb:47]: Are there any issues changing this token format?

File: app/models/computacenter/responsible_body_urns.rb [hunk 2]
```diff
@@ -48,6 +48,9 @@ def computacenter_identifier
         else
           "t#{companies_house_number.to_i}"
         end
+      when 'FurtherEducationCollege'
+        # potential n+1
+        schools.first&.ukprn
```
Comment (jamesrose) [inline app/models/computacenter/responsible_body_urns.rb:53]: Is it worth looking at whether we're eager loading schools where we use this method?

File: app/models/preorder_information.rb [hunk 3]
```diff
@@ -41,15 +41,15 @@ def infer_status
     elsif school_will_order_devices? && any_school_users? && chromebook_information_complete?
       if school.can_order_devices_right_now?
         'school_can_order'
-      elsif (school.std_device_allocation&.devices_ordered || 0).positive?
+      elsif (school.std_device_allocation&.devices_ordered || 0).positive? # bug should do all device types
```
Comment (jamesrose) [inline app/models/preorder_information.rb:44]: Good spot! Please could you add a Trello card if you don't want to fix it in this PR?

File: app/models/user.rb [hunk 4]
```diff
@@ -146,6 +146,14 @@ def effective_responsible_bodies
     user_schools.map { |us| us.school&.responsible_body }.prepend(responsible_body).compact.uniq
   end
 
+  def computacenter_account_orgs
+    effective_responsible_bodies + further_education_schools
+  end
+
+  def further_education_schools
+    user_schools.map(&:school).filter { |s| s.is_a?(FurtherEducationSchool) }
```
Comment (jamesrose) [inline app/models/user.rb:154]: Can we do this with AR?
```suggestion
    schools.where(type: 'FurtherEducationSchool')
```

File: app/services/importers/fe_csv.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          email_address: row['Contact email'],
+          full_name: row['Contact name'],
+          role: 'contact',
+          title: nil,
+          phone_number: row['Contact phone'],
+        )
+
+        next if school.preorder_information.present?
+
+        school.create_preorder_information!(
```
Comment (jamesrose) [inline app/services/importers/fe_csv.rb:66]: Do we need to prevent FE colleges from being able to modify this in the UI?

**Review by jamesrose — COMMENTED** [review]
@asmega Great work! I've had a first pass over this with some questions.

File: app/models/preorder_information.rb (hunk 3 above)
Comment (asmega) [reply app/models/preorder_information.rb:44]: I've raised a ticket for this to keep track of it

File: app/form_objects/school_search_form.rb (hunk 1 above)
Comment (asmega) [reply app/form_objects/school_search_form.rb]: I've changed this back, although not technically correct I think it's the safest thing to do

File: app/models/computacenter/responsible_body_urns.rb (hunk 2 above)
Comment (asmega) [reply app/models/computacenter/responsible_body_urns.rb:53]: I had a look for uses of `computacenter_identifier` and most seem to be referring to one RB or if it is a list it's a short list. The only exception is the "closed schools" in the `/computacenter` area. Given this and there are not many FE bodies for the moment I think this is safe for now

File: app/services/importers/fe_csv.rb (hunk 5 above)
Comment (asmega) [reply app/services/importers/fe_csv.rb:66]: None of the frontend changes have been done yet. But yes if we want to keep everything on the RB we can remove it from the interface. If we want to do this we may also want to remove some other features such as being able to add users on the school.

Maybe `/schools/123456` will link to the current controller and `/schools/12345678` can link of to another controller and set of views

File: app/models/computacenter/responsible_body_urns.rb (hunk 2 above)
Comment (asmega) [reply app/models/computacenter/responsible_body_urns.rb:53]: the "closed schools" area has been removed now so don't think this should cause an issue at least for now

File: app/models/user.rb (hunk 4 above)
Comment (asmega) [reply app/models/user.rb:154]: Even better, this turned out to be dead code probably when we switched to having the placeholder RB so I will delete

File: app/models/school.rb [hunk 6]
```diff
@@ -2,7 +2,8 @@ class School < ApplicationRecord
   has_paper_trail
   include PgSearch::Model
 
-  belongs_to :responsible_body
+  belongs_to :responsible_body, optional: true
```
Comment (jamesrose) [inline app/models/school.rb:5]: I don't think we need to make this optional anymore

**Review by jamesrose — APPROVED** [review]
Looks great. Thanks for your work on this Phil 💯 🚀

File: app/models/school.rb (hunk 6 above)
Comment (asmega) [reply app/models/school.rb]: good spot! I've removed this now

**Review by jamesrose — APPROVED** [review]

**Pull request merged**